### ISSUE TITLE
Bypass proxy when downloading MMS

### ISSUE THREAD

**Issue by NTICompass** [signalapp/Signal-Android#5721]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

---
### Bug description

I have a Virgin Mobile phone and MMS messages are not downloading.  It looks like it's timing out when downloading.  I did some tests and discovered that the timeout was due to the proxy.

On my PC, I tried to access `http://sobmmsc.vmobl.com:8088/ammsc/?xxxxxx` (with the `X-MDN` header correctly set) via the proxy and it timed out.  When I tried without the proxy, I was able to download the MMS message.

I've tried with both `68.28.31.7:80` and `205.239.233.136:81` as proxies (on my PC).

Is there a way to tell Signal to _bypass_ the proxy and download the MMS directly?  I've tried toggling the "wifi calling" option and that did nothing.  I also tried on both wifi and 3g and still got the timeout.
### Device info

**Device:** LG Volt (LGLS740) - Virgin Mobile
**Android version:** 6.0.1  - CyanogenMod 13.0 (From: http://androidforums.com/threads/just-throwing-this-out-there-cm13-for-g2-mini-sources.977839/page-17#post-7367175)
**Signal version:** 3.19.1 (beta from Play Store)
### Snippet from debug log

```
09-26 10:13:44.526 28380 28380 W ConversationItem: Content location: http://sobmmsc.vmobl.com:8088/ammsc/?xxxxxx
09-26 10:13:44.601 28380 28399 W MmsDownloadJob: Downloading mms at sobmmsc.vmobl.com
09-26 10:13:44.603 28380 28399 W CompatMmsConnection: Falling back to try receiving via Lollipop API
09-26 10:13:44.610 28380 28399 W IncomingLollipopMmsConnection: downloading multimedia from http://sobmmsc.vmobl.com:8088/ammsc/?xxxxxx to content://org.thoughtcrime.provider.securesms.mms/mms/1474899224609
09-26 10:13:44.655 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.662 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.689 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.695 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.714 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.720 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.727 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:13:44.733 28380 28380 W PhoneNumberFormatter: Got local CC: US
09-26 10:14:33.329 28380 30672 W WebSocketConnection: Sending keep alive...
09-26 10:14:33.606 28380 30671 W OkHttpClientWrapper: onMessage: BINARY
09-26 10:14:33.612 28380 30671 W WebSocketConnection: WSC onMessage()
09-26 10:14:33.619 28380 30671 W WebSocketConnection: Message Type: 2
09-26 10:14:37.678 28380 28405 W MessageRetrievalService: Application level read timeout...
09-26 10:14:37.681 28380 28405 W MessageRetrievalService: Network requirement: true, active activities: 1, push pending: 0
09-26 10:14:37.684 28380 28405 W MessageRetrievalService: Reading message...
09-26 10:14:44.473 28380 28380 W PassphraseRequiredActionBarActivity: onPause()
09-26 10:14:44.513 28380 28380 W KeyCachingService: onStartCommand, org.thoughtcrime.securesms.service.action.ACTIVITY_STOP_EVENT
09-26 10:14:44.513 28380 28380 W KeyCachingService: Decrementing activity count...
09-26 10:14:44.513 28380 28380 W KeyCachingService: Starting timeout: 1800000
09-26 10:14:44.515 28380 28380 W MessageRetrievalService: Active Count: 0
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection: java.util.concurrent.TimeoutException: timeout when waiting for MMS
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.thoughtcrime.securesms.mms.LollipopMmsConnection.waitForResult(LollipopMmsConnection.java:74)
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.thoughtcrime.securesms.mms.IncomingLollipopMmsConnection.retrieve(IncomingLollipopMmsConnection.java:86)
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.thoughtcrime.securesms.mms.CompatMmsConnection.retrieve(CompatMmsConnection.java:74)
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.thoughtcrime.securesms.jobs.MmsDownloadJob.onRun(MmsDownloadJob.java:97)
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.thoughtcrime.securesms.jobs.MasterSecretJob.onRun(MasterSecretJob.java:18)
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.whispersystems.jobqueue.JobConsumer.runJob(JobConsumer.java:76)
09-26 10:14:44.648 28380 28399 W IncomingLollipopMmsConnection:     at org.whispersystems.jobqueue.JobConsumer.run(JobConsumer.java:46)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob: ws.com.google.android.mms.MmsException: java.util.concurrent.TimeoutException: timeout when waiting for MMS
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.thoughtcrime.securesms.mms.IncomingLollipopMmsConnection.retrieve(IncomingLollipopMmsConnection.java:97)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.thoughtcrime.securesms.mms.CompatMmsConnection.retrieve(CompatMmsConnection.java:74)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.thoughtcrime.securesms.jobs.MmsDownloadJob.onRun(MmsDownloadJob.java:97)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.thoughtcrime.securesms.jobs.MasterSecretJob.onRun(MasterSecretJob.java:18)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.whispersystems.jobqueue.JobConsumer.runJob(JobConsumer.java:76)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.whispersystems.jobqueue.JobConsumer.run(JobConsumer.java:46)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob: Caused by: java.util.concurrent.TimeoutException: timeout when waiting for MMS
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.thoughtcrime.securesms.mms.LollipopMmsConnection.waitForResult(LollipopMmsConnection.java:74)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    at org.thoughtcrime.securesms.mms.IncomingLollipopMmsConnection.retrieve(IncomingLollipopMmsConnection.java:86)
09-26 10:14:44.663 28380 28399 W MmsDownloadJob:    ... 5 more
09-26 10:15:40.542 28380 28380 W PassphraseRequiredActionBarActivity: onResume()
```
### Link to MmsMonster log with some interesting info

https://gist.github.com/anonymous/931fc2507d8c519bb3fed1b1ef949963

Snippet:

```
 1025 1249 W OutgoingMmsConnection: org.apache.http.conn.HttpHostConnectException: Connection to http://68.28.31.7:80 refused
```

**Comment by E3V3A** [comment]
MMS is another terrible security risk. There are loads of MMS attacks out there, so MMS should be avoided at all costs. In fact on Androids you should disable all use of MMS. In qddition you wanna bypass any proxies... Are you kidding!? 

Please close issue as N/A or rejected.

**Comment by moxie0** [comment]
@E3V3A That comment isn't helpful, please only post here if you have new information to add. This is a legitimate issue.

**Comment by E3V3A** [comment]
Sorry, you are right. I just thought it sounded weird to try to circumvent the proxy. It seem that *Virgin Mobile* simply doesn't want you to use a proxy, for obvious reasons. Thus I would rather *not* be able to download an MMS than using their servers directly.

**Comment by NTICompass** [comment]
@E3V3A: Pretty sure both SMS and MMS are a security risk.  As I was testing downloading MMS messages on my PC, I noticed that they are downloaded via `http` (not `https`) and the only "authentication" is the `X-MDN` header.  Also, aren't the proxy servers over `http`, also?

I did kinda wonder if I could edit the proxy to route through my network then "edit" the MMS messages before the phone downloads them... but that's a different thing.

Anyway, maybe disabling the MMS proxy isn't a safe option, but it seems to be the only way to download MMS messages as of now.  At least until I can find another proxy server to use (Virgin Mobile or otherwise).
